Docker environment

binakot opened the pull request:
* Added Docker environment.
* Replace SQLite with PostgreSQL.

**NOT READY FOR MERGE YET**

**NEED HELP**

As it was said in the issue #1, there are no docs for install. Probably I have an error somewhere. @hexchan, please check my PR. When I run the server I got an error:

```text
Execute generators
Target dirs: []
Images found: 0
Traceback (most recent call last):
  File "generators/partymaker.py", line 227, in <module>
    available_images = load_json_from_file(available_images_filename)
  File "generators/partymaker.py", line 76, in load_json_from_file
    with path.open(mode='r') as f:
  File "/usr/local/lib/python3.7/pathlib.py", line 1186, in open
    opener=self._opener)
  File "/usr/local/lib/python3.7/pathlib.py", line 1039, in _opener
    return self._accessor.open(self, flags, mode)
FileNotFoundError: [Errno 2] No such file or directory: '/app/generators/../dev/upload_fakes/images.json'
Build frontend
Build started at: Fri Apr 12 21:52:16 UTC 2019
Prepare output folder
Build scripts

src/frontend/main.js → dev/frontend/scripts.js...
created dev/frontend/scripts.js in 99ms
Build styles
Copy assets
Copy libraries
Done
Apply database migration
Operations to perform:
  Apply all migrations: admin, auth, captcha, contenttypes, imageboard, moderation, sessions
Running migrations:
  Applying contenttypes.0001_initial... OK
  Applying auth.0001_initial... OK
  Applying admin.0001_initial... OK
  Applying admin.0002_logentry_remove_auto_add... OK
  Applying admin.0003_logentry_add_action_flag_choices... OK
  Applying contenttypes.0002_remove_content_type_name... OK
  Applying auth.0002_alter_permission_name_max_length... OK
  Applying auth.0003_alter_user_email_max_length... OK
  Applying auth.0004_alter_user_username_opts... OK
  Applying auth.0005_alter_user_last_login_null... OK
  Applying auth.0006_require_contenttypes_0002... OK
  Applying auth.0007_alter_validators_add_error_messages... OK
  Applying auth.0008_alter_user_username_max_length... OK
  Applying auth.0009_alter_user_last_name_max_length... OK
  Applying captcha.0001_initial... OK
  Applying captcha.0002_auto_20190126_1145... OK
  Applying captcha.0003_auto_20190129_1529... OK
  Applying imageboard.0001_initial... OK
  Applying imageboard.0002_auto_20180810_0800... OK
  Applying imageboard.0003_remove_thread_last_post_hid... OK
  Applying imageboard.0004_board_custom_style... OK
  Applying imageboard.0005_auto_20181010_1254... OK
  Applying imageboard.0006_auto_20181225_1051... OK
  Applying imageboard.0007_auto_20190111_1136... OK
  Applying imageboard.0008_post_session_id... OK
  Applying imageboard.0009_auto_20190129_1529... OK
  Applying imageboard.0010_auto_20190204_1314... OK
  Applying imageboard.0011_auto_20190208_1149... OK
  Applying imageboard.0012_auto_20190208_1153... OK
  Applying moderation.0001_initial... OK
  Applying moderation.0002_auto_20190127_1020... OK
  Applying moderation.0003_auto_20190127_1155... OK
  Applying moderation.0004_auto_20190127_1155... OK
  Applying moderation.0005_auto_20190129_1529... OK
  Applying sessions.0001_initial... OK
Starting server
[12/Apr/2019 21:52:26] "GET / HTTP/1.1" 200 2145
[12/Apr/2019 21:52:27] "GET /static/style.css HTTP/1.1" 200 16879
[12/Apr/2019 21:52:27] "GET /static/scripts.js HTTP/1.1" 200 15655
[12/Apr/2019 21:52:27] "GET /static/libs/lightbox/css/lightbox.css HTTP/1.1" 200 3714
[12/Apr/2019 21:52:27] "GET /static/libs/underscore-min.js HTTP/1.1" 200 18069
[12/Apr/2019 21:52:27] "GET /static/libs/lightbox/js/lightbox.min.js HTTP/1.1" 200 9418
[12/Apr/2019 21:52:27] "GET /static/libs/jquery.min.js HTTP/1.1" 200 86927
[12/Apr/2019 21:52:27] "GET /static/libs/js.cookie.js HTTP/1.1" 200 3886
[12/Apr/2019 21:52:27] "GET /static/images/squared_metal.png HTTP/1.1" 200 16789
[12/Apr/2019 21:52:27] "GET /static/style.css.map HTTP/1.1" 200 8379
[12/Apr/2019 21:52:27] "GET /static/images/hexchan.svg HTTP/1.1" 200 2608
[12/Apr/2019 21:52:27] "GET /static/libs/lightbox/js/lightbox.min.map HTTP/1.1" 200 9672
[12/Apr/2019 21:52:27] "GET /static/scripts.js.map HTTP/1.1" 200 28998
[12/Apr/2019 21:52:27] "GET /static/libs/lightbox/images/loading.gif HTTP/1.1" 200 8476
[12/Apr/2019 21:52:27] "GET /static/libs/lightbox/images/next.png HTTP/1.1" 200 1350
[12/Apr/2019 21:52:27] "GET /static/libs/lightbox/images/prev.png HTTP/1.1" 200 1360
[12/Apr/2019 21:52:27] "GET /static/libs/lightbox/images/close.png HTTP/1.1" 200 280
[12/Apr/2019 21:52:27] "GET /static/favicon.png HTTP/1.1" 200 1218
Internal Server Error: /captcha/
```

But server is running. On localhost:8000 I see the empty page:

<img width="1280" alt="Снимок экрана 2019-04-13 в 1 03 35" src="https://user-images.githubusercontent.com/1783893/56068571-06a62d00-5d88-11e9-9dfc-556223ef71e3.png">

As I can see the reason of this problem is captcha:

<img width="665" alt="Снимок экрана 2019-04-13 в 1 04 38" src="https://user-images.githubusercontent.com/1783893/56068634-3a815280-5d88-11e9-9262-40baa6e52d80.png">

How can I fix that? What is the reason?

binakot commented:
I just need to know what I have to do when run the app here:
https://github.com/hexchan/hexchan-engine/pull/2/files#diff-60713fb19f6c50a2473835df63ff17f9

hexchan commented:
Hi. I dunno anything about Docker, but I'll try to help.

Before running `partymaker.py` you have to run `imagemaker.py` and point it to some directory with images as the first argument. It'll create symlinks to those images inside `dev/upload_fakes/images` dir, generate thumbnails in `dev/upload_fakes/thumbs` and save all image metadata in `dev/upload_fakes/images.json`.

`Images.json` file is used in `partymaker.py` to pick random images for generated user posts. Any images could be picked multiple times and a symlink is created with a new name in `dev/uploads`. No file copies ever get created.

binakot commented:
I just updated the docker env based on new INSTALL.md file.
But CI is broken now because of error:

```bash
...
django.db.utils.OperationalError: could not translate host name "db" to address: Name or service not known
The command "./manage.py test" exited with 1.
```

I changed the database connection settings to dockerized postgres with domain name `db`. Is it possible to specify multiple environments and provide it before run the app? For example, put the variable or smth else to `migrate` script.

Btw this is the screenshots of hexchan with fake content running inside docker containers:

<img width="1279" alt="Снимок экрана 2019-05-04 в 1 49 39" src="https://user-images.githubusercontent.com/1783893/57169578-3d63e600-6e10-11e9-835c-286fb58a2ba5.png">

<img width="1278" alt="Снимок экрана 2019-05-04 в 1 49 47" src="https://user-images.githubusercontent.com/1783893/57169579-3fc64000-6e10-11e9-92ca-95509baa2e49.png">

binakot commented:
Work is done! `Docker-compose` & `Docker stack` are ready to work :octocat: 🐙
